src/app/projects/projlist-map/projlist-map.component.ts

```diff
@@ -301,33 +259,49 @@ export class ProjlistMapComponent implements AfterViewInit, OnChanges, OnDestroy
 
   /**
     * Removes deleted / draws added projects.
+    *
+    * @param   {Project[]} deletedApps Deleted projects.
+    * @param   {Project[]} addedApps Added projects.
+    * @returns {void}
     */
-  private drawMap(deletedApps: Project[], addedApps: Project[]) {
-    // console.log('drawing map');
-    // console.log('deleted =', this.deleted);
-    // console.log('added =', this.added);
-
+  private drawMap(deletedApps: Project[], addedApps: Project[]): void {
     // remove deleted apps from list and map
     deletedApps.forEach(app => {
       const markerIndex = _.findIndex(this.markerList, { projectId: app._id });
+      const shapefileIndex = _.findIndex(this.shapefileList, { projectId: app._id });
       if (markerIndex >= 0) {
         const markers = this.markerList.splice(markerIndex, 1);
         this.markerClusterGroup.removeLayer(markers[0]);
       }
+      if (shapefileIndex >= 0) {
+        this.shapefileList.splice(shapefileIndex, 1);
+      }
     });
 
     // draw added apps
     addedApps.forEach(app => {
-      // add marker
-      if (app.centroid.length === 2) { // safety check
-        const title = `${app.name}\n`
+      // If there is a shapefile for one of the projects, display it instead of a pin.
+      if (app.shapefiles && app.shapefiles.length > 0) {
+        app.shapefiles.forEach(projectShapefile => {
+          const escapedName = encode(projectShapefile.documentFileName).replace(/\(/g, '%28').replace(/\)/g, '%29').replace(/\\/g, '_').replace(/\//g, '_').replace(/\%2F/g, '_');
```

Comment:
It feels like this could be done a more efficient way using regex

Reply:
So like a single "replace" call that captures each pattern? We'd still need to replace a pattern with a specific thing(for example, "\\" needs to get replaced with "_"), so I don't know how we could do that more succinctly.